gcc/rust/backend/rust-compile-intrinsic.cc

```diff
@@ -837,6 +929,118 @@ atomic_load_handler_inner (Context *ctx, TyTy::FnType *fntype, int ordering)
   return fndecl;
 }
 
+static tree
+atomic_compare_exchange_handler_inner (Context *ctx, TyTy::FnType *fntype,
+				       tree weak_value,
+				       int success_memorder_raw,
+				       int failure_memorder_raw)
+{
+  rust_assert (fntype->get_params ().size () == 3);
+  rust_assert (fntype->get_num_substitutions () == 1);
+
+  tree lookup = NULL_TREE;
+  if (check_for_cached_intrinsic (ctx, fntype, &lookup))
+    return lookup;
+
+  auto fndecl = compile_intrinsic_function (ctx, fntype);
+
+  std::vector<Bvariable *> param_vars;
+  std::vector<tree> types;
+  compile_fn_params (ctx, fntype, fndecl, &param_vars, &types);
+
+  if (!ctx->get_backend ()->function_set_parameters (fndecl, param_vars))
+    return error_mark_node;
+
+  enter_intrinsic_block (ctx, fndecl);
+
+  auto dst = ctx->get_backend ()->var_expression (param_vars[0], Location ());
+  auto old = ctx->get_backend ()->var_expression (param_vars[1], Location ());
+  auto src = ctx->get_backend ()->var_expression (param_vars[2], Location ());
+  auto success_memorder = make_unsigned_long_tree (ctx, success_memorder_raw);
+  auto failure_memorder = make_unsigned_long_tree (ctx, failure_memorder_raw);
+
+  TREE_READONLY (dst) = 0;
+
+  auto monomorphized_type
+    = fntype->get_substs ()[0].get_param_ty ()->resolve ();
+  auto builtin_name
+    = build_atomic_builtin_name ("atomic_compare_exchange_",
+				 fntype->get_locus (), monomorphized_type);
+
+  auto atomic_cmp_xchg_raw = NULL_TREE;
+  BuiltinsContext::get ().lookup_simple_builtin (builtin_name,
+						 &atomic_cmp_xchg_raw);
+  rust_assert (atomic_cmp_xchg_raw);
+
+  auto atomic_cmp_xchg = build_fold_addr_expr_loc (Location ().gcc_location (),
+						   atomic_cmp_xchg_raw);
+
+  // The gcc builtin has a fun behavior where if *ptr and *expected are not
+  // equal, it will write *ptr into *expected. So we actually need to create a
+  // tmp variable which is copy(old) and pass the adress as an argument to the
+  // builtin. We then perform the call, and return {tmp, result} as the old
+  // value will have been written into tmp
+
+  // let tmp = old;
+  // let tmp_addr = &tmp;
+  auto tmp_init = NULL_TREE;
```

Comment:
You are missing:

```
ctx->add_statement (tmp_init);
```

The go backend interface gives us a temp variable but you need to add it to the current block.